site/content/en/development/design-docs/multiple-slb.md

```diff
@@ -0,0 +1,38 @@
+
+However, there are scenarios that we need multiple SLBs in the cluster. For example, users could set different outbound IPs for different agent pools. In addition, more services are supported with multiple SLBs because there is a 4Mb restriction in an ARM request body.
+
+Basically, we would like to support the mode selection annotation (`service.beta.kubernetes.io/azure-load-balancer-mode`) with SLB, similar to what it is with basic LB.
+
+## User experience
+
+Currently, only one SLB is allowed in one cluster. If users set the mode selection annotation when using the SLB, he would get an error. After this feature implemented, the users could choose if they want a single SLB (current behavior) or they want multiple SLBs, with one SLB binding to one agent pool. If the annotation is not set, everything is unchanged: every agent node in the cluster would join the backend pool of the SLB and only one SLB is allowed. If the annotation is set to `__auto__` or `agentpool1Name,agentpool2Name`, there would be a 1: 1 mapping between the SLBs and the agent pools. Specifically, for `__auto__` mode, the SLB with the minimum load balancing rules would be selected to serve the newly created service; for`agentpool1Name,agentpool2Name` mode, the SLB of the given agent pools with the minimum load balancing rules would be selected. In this way, users could set different outbound rules for different agent pools. If there is no SLB of the given agent pools, a new SLB would be created.
+
+## Implementation
+
+There are three selection mode of the SLB: `default`, `__auto__` and `agentpool1,agentpool2`. If the annotation `service.beta.kubernetes.io/azure-load-balancer-mode` is not set for the SLB cluster, the behavior would be the same as it is now. If it is set to `__auto__`, new services should choose the SLB with minimum rules. If it is set to `agentpool1,agentpool2`, new services should choose the SLB with minimum rules of the given agent pools.
+
+The following are the implementations in detail.
+
+1. Introduce a new cloud provider config `enableMultipleStandardLoadBalancers`, default to `false`. When `enableMultipleStandardLoadBalancers=false`, the mode selection annotation `service.beta.kubernetes.io/azure-load-balancer-mode` would be ignored when using the SLB.
+
+2. When `enableMultipleStandardLoadBalancers=true`:
+
+    - If the annotation `service.beta.kubernetes.io/azure-load-balancer-mode` is not set, the SLB of the primary agent pool would be selected for the services.
+
+    - If `service.beta.kubernetes.io/azure-load-balancer-mode = __auto__`, the SLB with minimum rules would be selected for the services.
+
+    - If `service.beta.kubernetes.io/azure-load-balancer-mode = agentpoolName1,agentpoolName2`, the SLB of the given agent pools with minimum rules would be selected for the services. If there is no corresponding SLB, the cloud provider would start to create one with the naming format `agentpoolName` or `agentpoolName-internal`.
```

Comment:
As far as I know, there is no native concept of "agent pool" in k8s / the Azure cloud provider. That's an AKS Engine / AKS concept. The current [documentation](https://kubernetes-sigs.github.io/cloud-provider-azure/topics/loadbalancer/#load-balancer-selection-modes) only mentions `availability sets`. I think we should consider grouping by availability set or by scale set since not every Azure provisioner will have a concept of "agent pool". For example, Cluster API has MachinePools and MachineDeployments but no agent pools.

Reply:
agreed, we need rename this to VMSS/VMAS